### ISSUE TITLE
Incorrect TCH001 when used by dataclass

### ISSUE THREAD

**Issue by nstarman** [charliermarsh/ruff#2665]
In the following example ruff flags ``Bar`` for TCH001, but python requires `Bar` to be run-time accessible.

```python

from bar import Bar

@dataclass
class Foo:
    x: Bar
```

**Comment by charliermarsh** [comment]
(The original plugin also flags this as an error.)

**Comment by charliermarsh** [comment]
Wow, fascinating. The annotations are evaluated IFF in a module or class scope: https://docs.python.org/3/reference/simple_stmts.html#annotated-assignment-statements.

**Comment by charliermarsh** [comment]
Very fun, will fix!

https://twitter.com/charliermarsh/status/1623389778375843840

**Comment by nstarman** [comment]
Thanks for the quick fix!

**Comment by charliermarsh** [comment]
Are you able to lint the file that this occurred on? Just so I can test before this release.

**Comment by nstarman** [comment]
Ahh. Private repo. Let me make an example real quick... Ok this definitely tripped TCH001 and then broke the dataclass when I fixed it.

```python
from dataclasses import dataclass
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Mapping

@dataclass
class Test:
    arg: Mapping[str, str]
```

**Comment by charliermarsh** [comment]
Ok cool! That now yields `TCH004`, telling you to take it out of the block. And if you take it out of the block, no errors are triggered, so I think we're good :) Thanks.
